mllib/src/main/scala/org/apache/spark/ml/regression/RobustRegression.scala

```diff
@@ -0,0 +1,497 @@
+   */
+  @Since("2.1.0")
+  def setStandardization(value: Boolean): this.type = set(standardization, value)
+  setDefault(standardization -> true)
+
+  /**
+   * Sets the maximum number of iterations.
+   * Default is 100.
+   * @group setParam
+   */
+  @Since("2.1.0")
+  def setMaxIter(value: Int): this.type = set(maxIter, value)
+  setDefault(maxIter -> 100)
+
+  /**
+   * Sets the convergence tolerance of iterations.
+   * Smaller value will lead to higher accuracy with the cost of more iterations.
+   * Default is 1E-6.
+   * @group setParam
+   */
+  @Since("2.1.0")
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-6)
+
+  /**
+   * Whether to over-/under-sample training instances according to the given weights in weightCol.
+   * If not set or empty, all instances are treated equally (weight 1.0).
+   * Default is not set, so all instances have weight one.
+   * @group setParam
+   */
+  @Since("2.1.0")
+  def setWeightCol(value: String): this.type = set(weightCol, value)
+
+  override protected def train(dataset: Dataset[_]): RobustRegressionModel = {
+    val numFeatures = dataset.select(col($(featuresCol))).first().getAs[Vector](0).size
+    val w = if (!isDefined(weightCol) || $(weightCol).isEmpty) lit(1.0) else col($(weightCol))
+
+    val sqlContext = dataset.sqlContext
+    import sqlContext.implicits._
+    val instances: RDD[Instance] = dataset.select(col($(labelCol)).as("label"),
+      w.as("weight"), col($(featuresCol)).as("features")).as[Instance].rdd
+
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)
+
+    val featuresSummarizer = {
+      val seqOp = (c: MultivariateOnlineSummarizer, v: Instance) => c.add(v.features, v.weight)
+      val combOp = (c1: MultivariateOnlineSummarizer, c2: MultivariateOnlineSummarizer) => {
+        c1.merge(c2)
+      }
+      instances.treeAggregate(new MultivariateOnlineSummarizer)(seqOp, combOp)
+    }
+
+    val featuresStd = featuresSummarizer.variance.toArray.map(math.sqrt)
+    val regParamL2 = $(regParam)
+
+    val costFun = new HuberCostFun(instances, $(fitIntercept), $(standardization),
+      featuresStd, regParamL2, $(m))
+
+    /**
+     * Huber loss function should be optimized by LBFGS-B, but breeze LBFGS-B has bug and
+     * can't be used currently. We figure out one work around with modified LBFGS.
+     * Since we known huber loss function is convex in space "\sigma > 0" and the bound
+     * "\sigma = 0" is unreachable. The solution of loss function will not be on the bound.
+     * We still optimize the loss function by LBFGS but limit the step size when doing line search
+     * of each iteration. We should verify the step size generated by each line search in the
+     * space "\sigma > 0".
+     */
+    val optimizer = new BreezeLBFGS[BDV[Double]]($(maxIter), 10, $(tol)) {
+      override protected def determineStepSize(
```

Comment:
Update LBFGS-B as already fix bug in https://github.com/scalanlp/breeze/pull/633